[MIG] default_warehouse_from_sale_team: Migrate to 15.0

rolandojduartem opened the pull request:
Module default_warehouse_from_sale_team migration to 15.0

In this module I added some fixes to become it functional at 15.0.
- Now, the page's name on crm_team_view_form is "members_users". Old reference was "members"
- Fix _search_rule inheritance adding packaging_id argument. original method URL: https://github.com/odoo/odoo/blob/15.0/addons/stock/models/stock_rule.py#L465

Also, as part of migration to 15.0 of default_warehouse_from_sale_team and
default_warehouse_from_sale_team_jit, the dependency procurement_jit was removed
(https://github.com/odoo/odoo/commit/a838f2af683e03cd03ba02532f002ccd5c2642f5). This change caused
that default_warehouse_from_sale_team_jit
(https://github.com/Vauxoo/addons-vauxoo/tree/14.0/default_warehouse_from_sale_team_jit)
be deprecated on Odoo v15.0 and the main function of that model can be added in
default_warehouse_from_sale_team.

Technical tests are passing!

nhomar committed:
[IMP] Better test naming, and cleanup of test (#883)

sale_team_warehouse => default_warehouse_from_sale_team
* Better test naming.
* [IMP] Better module naming.

nhomar committed:
8.0 warehouse default (#889): Test was incorrectly invoqued

JulioSerna committed:
default_warehouse_from_sale_team: new version (#1032) (#1281)

This module formerly just add a default_warehouse field to the sales
    team model that is used to set the `warehouse_id` field in the sale order
    when the current user has defined a `Default Sale Team`

    We extend this functionality by:

    - Improve consistency: Add a constraint that do not let us to add a
      default sale team if the user do not belongs to the sales team.
    - Improve usability: Add an automated action/server action to update the
      user default sales team every time that a sales teams by:

        - add user's default sale team if empty.
        - remove default sale team from user if not longer in sale team.
        - dummy write to update m2m in users to make filtering rules work.

    - Extend scope: Now both stock picking and stock.picking type models
      also have default warehouse_id functionality taking into account the
      user default sale team.
    - Compatibility fix: in order to be compatible with one of our customers
      instances we remove the _auto = False for WarehouseDefault api.

    - New feature for Permissions and Security: Taking advantage of the
      default_warehouse field in the sales team model now we can filter the
      records (picking type and picking model) to only show those
      records that match with the user sale teams default_warehouse. To
      accomplish this I:

        - add new groups to manage the records access by user:

          * Default Warehouse / Limited access to pickings (filtered by sale teams)
          * Default Warehouse / Limited access to stock pickings (filtered by sale teams)
          * Default Warehouse / Access to all picking types (filtered by sale teams)
          * Default Warehouse / Access to all pickings (filtered by sale teams)

        - add new m2m field in the res.user model used in the new ir.rules.
          this onw is showed as a readonly field (only informative) to know
          the warehouse of the user.
        - add new ir.rule records, one for each default warehouse group. This
          one will let us to only show the records for the current user sale
          teams default_warehouse or to do not take into account the sale teams
          and show all the records to the user.

    Module version, description and Spanish translations were updated: missing
    translation template was added. Also the module was modified to follow
    OCA guideline:

    - models in one model file per model/module.

    default_warehouse_from_sale_team: defined 'picking_type_id' default as well from the sale team werahouse

    default_warehouse_from_sale_team: add test to validate pick type from sale team

    default_warehouse_from_sale_team: return limit 1 picking type by defualt to avoid 'Expected singleton' when create record

    default_warehouse_sale_team: add docstring to method

    default_warehouse_from_sale_team: Allow to management that users can see only the journals that match with the sale team

    default_warehouse_from_sale_team: deleted file not used

    apply pylint

    apply pylint

    default_warehouse_from_sale_team: propagate SUPERUSER_ID in read (#1042)

    * [IMP] default_warehouse_from_sale_team: propagate SUPERUSER_ID in read
    when user create from sale pickings chained in another warehouse without
    access to read.
    * [IMP] default_warehouse_from_sale_team: adapt to old and new api read
    * [IMP] default_warehouse_from_sale_team: adapt read to use has_group to
    validate in what user required SUPERUSER_ID.

    default_warehouse_from_sale_team: added rule to allow to see own warehouse by sale team

    default_warehouse_from_sale_team: Adding test unit for validation in policy from sale (#1064)

    * [IMP] default_warehouse_from_sale_team: add journal stock from sale team
    in journal entries of stock valuation
    * [ADD] validation police from sale order defined in sale team warehouse

    default_warehouse_from_sale_team: Move fields default_warehouse… (#1066)

    * [REF] default_warehouse_from_sale_team: Move fields default_warehouse and journal_team_ids to new page Settings warehouse from sale team.

    * [REF] default_warehouse_from_sale_team: Move field journal_stock_id to new page Settings warehouse from sale team.

    * [ADD] default_warehouse_from_sale_team: Adapt traslation.

    * [FIX] default_warehouse_from_sale_team: Fix pylint.

    default_warehouse_from_sale_team: Change field journal_team_ids to one2many. [ADD] default_warehouse_from_sale_team: Add field section_id in account_invoice model to reference sale teams.

    default_warehouse_from_sale_team: in order to allow search and see all record from fields remove selection widget to use origianl widget

    default_warehose_from_sale_team: Modify search of warehouse from the move

    Adding stock_picking_type.warehouse_id field to generate the policy while
    taking into consideration, the warehouse

    default_warehouse_from_sale_team: Modified view_picking_type_form view

    Adding attribute required in warehouse field

    default_warehouse_from_sale_team: adapt class to allow get sequence in base to sale tema in models like: * sale.order * purchase.order * purchase.requisition

    default_warehouse_from_sale_team: change data demo from data to demo

    addons-vauxoo: add disable message by duplicate methods

    default_warehouse_from_sale_team: Modified updated dictionary from super

    Modified updated dictionary from super method because warehouse_id default

    default_warehouse_from_sale_team: Modified updated dictionary from super

    Revert modified updated dictionary from super method because warehouse_id
    default

    default_warehouse_from_sale_team: add priority to get warehouse from warehouse of picking type of stock move because warehouse_id of stock move can be another different by stock move propagated like re-supply move from another warehouse

    default_warehouse_from_sale_team: get picking type from picking of move

    [IMP] default_warehouse_from_sale_team: migration to version11.0

[IMP] default_warehouse_from_sales_team: remove file unnecessary

[IMP] default_warehouse_from_sales_team: remove method read unnecessary

deivislaya committed:
[FIX] default_warehouse_from_sale_team: Update get_id method in sequence (#1290)

* [ADD] default_warehouse_from_sale_team: Update get_id method in sequence

* [ADD] default_warehouse_from_sale_team: Access Rule

JulioSerna committed:
[IMP] default_warehouse_from_sales_team: adapt method in order to avoid (#1294)

error by expected singleton

hugho-ad committed:
[MIG] default_warehouse_from_sale_team: Migration to 12.0

hugho-ad committed:
[REF] default_warehouse_from_sale_team: Default journal at invoice is get from the sale team

hugho-ad committed:
[FIX] Lints

hugho-ad committed:
[REF] i18n: Update translations

Commit:
[FIX] default_warehouse_from_sale_team: Wrong UoM Reference

CarmenMiranda committed:
[I18N] default_warehouse_from_sale_team: Update translations terms

By merging the two Spanish language translation files we are
solving the issue of duplicated terms on the translations of this
module.
Also changing some of the messages and translations of this module.
And increasing the module version for this major change.

CarmenMiranda committed:
[I18N] default_warehouse_from_sale_team: Typo in translation

And deleting a term that is not need it.

CarmenMiranda committed:
[FIX] default_warehouse_from_sale_team: Default Sales Team Constraint

When a user is created without a default sales team or if a user is
being deleted the field `sale_team_id` will be set to False, the
constraint will fail since the value False is not on the `sale_team_ids`
list.

luisg123v committed:
[FIX] default_warehouse_from_sale_team: Adapt tests to v12

Commit:
[IMP] default_warehouse_from_sale_team: Remove nested for on ir_rule

This change simplifies the code and avoids bad practices

luisg123v committed:
[FIX] default_warehouse_from_sale_team: Remove unnecessary headers

This removes:
- Hencoding headers, fixing lint `utf8-coding-comment`
- Headers like `# Part of Odoo`
- And other ones

luisg123v committed:
[MIG] default_warehouse_from_sale_team: Migrate to v13

The following is performed:
- Replace account invoice model by account move
- import openerp -> import odoo
- Fix class names
- Fix indentation in data files
- Adapt tests

Co-Authored-By: José Suniaga [Vauxoo] <josemiguel@vauxoo.com>

Commit:
[FIX] default_warehouse_from_sale_team: avoid sequence re-call

The sequence is calling two times, one from original model and
another from 'default.warehouse' inherit, this fix avoid the
re-call to next_sequence number if it was called previously,
this is just by checking if the name has been already set.

Before:
- Current Purchase Order P00039, next P000041.

Now:
- Current Purchase Order P00041, next P00042.

CarmenMiranda committed:
[FIX] default_warehouse_from_sale_team: CRM Team Read Permission

Using `sudo()` to avoid creating `ir.model.access` but allowing all the
users with access to create a purchase order using the default
warehouse.

luisg123v committed:
[MIG] default_warehouse_from_sale_team: Migrate to 14.0

Odoo now provides a native way to specify a default warehouse on the
user. So this module uses the same behavior, but giving priority if a
warehouse is set on the sales team.

Other tasks are also performed:
- Code and views are reorganized by model
- Models that were actually mixings are renamed to use the `mixing` suffix
- Test are improved to use the `Form` class
- Some Typos are fixed
- General code clean up

CarmenMiranda committed:
[IMP] default_warehouse_from_sale_team: Improving Rules

Adding to the rule `rule_default_warehouse_journal` an `or` if the
journal doesn't have its field `section_id` set.

Also, improving the domains of the rules to avoid list comprehensions
using the mapped for related fields.

luisg123v committed:
[REF] default_warehouse_from_sale_team: Sequence assignation in v14.0

Automatic sequence numbering was not working in 14.0. For instance, now
purchase requisitions have different sequence codes, depending on
their types.

luisg123v committed:
[IMP] default_warehouse_from_sale_team: Rename field Default Warehouse

Field name for default warehouse on sales teams was missing the proper
suffix on previous versions of this module.

This implements a migration script so the database column is renamed
when we're coming from previous versions.

luisg123v committed:
[FIX] default_warehouse_from_Sale_team: Don't overwrite team nor warehouse

When the partner is changed in a sale order, the current user's sales
team and warehouse should have priority over other ones.

luisg123v committed:
[FIX] default_warehouse_on_sale_team: typo in model names of mixins

mixing -> mixin

luisg123v committed:
[FIX] default_warehouse_from_sale_team: access error when chained rule

When a stock movement triggers an inventory rule that involves another
warehouse on which the current user has no access, an access error is
triggered. That's because, thanks to this module, users can only see
warehouses allowed in their sales teams.

To solve the above, access is granted to all warehouses when computing
inventory rules to apply.

luisg123v committed:
[IMP] default_warehouse_from_sale_team: Default journal for vendor bills

Currently, when creating a customer invoice, a default journal is set
if it's configured in the sales team. However, that feature only works
for customer invoices but not for vendor bills.

This change implements the default journal feature also for vendor
bills.

In addition, now company and currency are also considered when
computing default journal, for more accurate results.

desdelinux reviewed (commented):
@hugho-ad @luisg123v Should a pre-migration be created for v15?

cc @rolandojduartem

luisg123v commented:
What is this module being migrated for?

hugho-ad commented:
tanner

rolandojduartem commented:
Hi, @hugho-ad and @desdelinux, could you do the technical and functional review, please?

- Dummy MR: https://git.vauxoo.com/vauxoo/villagroup/-/merge_requests/167

luisg123v commented on default_warehouse_from_sale_team/README.rst:
```diff
@@ -0,0 +1,123 @@
+
+* Nhomar Hernández <nhomar@vauxoo.com> (Planner/Developer)
+* Katherine Zaoral <kathy@vauxoo.com> (Planner/Developer)
+* Luis González <lgonzalez@vauxoo.com> (Developer)
```
This is not really black, isort, prettier. Could you fix in the previous version, please? I think this is actually a bug

rolandojduartem replied:
Done!

luisg123v commented on default_warehouse_from_sale_team/__manifest__.py:
```diff
@@ -0,0 +1,33 @@
+    'category': 'Inventory/Inventory',
+    'version': '15.0.1.0.0',
+    'depends': [
+        'crm',
```
Why? This shouldn't be required

desdelinux replied:
@rolandojduartem you are right @luisg123v 

You can get to the sales teams menu from Sales / Configurations / Sales Teams

rolandojduartem replied:
Hi, @luisg123v and @desdelinux. Here I have a problem, when I install this module in a clean odoo, only these additional modules in the image (invoicing, purchase, inventory) are installed with default_warehouse and those modules do not have that configuration (Sales Teams). Do I add any dependencies to default_warehouse or keep it as before? I attach an image.

![Screenshot from 2022-03-04 10-28-04](https://user-images.githubusercontent.com/90422721/156802010-37dfd2fa-6604-453d-9c95-fca054e22c74.png)

desdelinux replied:
@luisg123v ¿añadimos Sales como dependencia?

luisg123v replied:
@rolandojduartem & @desdelinux,

Is the same behavior as in v14 (and previous versions). The sales module doesn't add the sales app, that's added by `sale_management`. I think we can keep it that way.

rolandojduartem replied:
Well! This change and the others have already been applied. @luisg123v 

cc @desdelinux

luisg123v commented on default_warehouse_from_sale_team/README.rst:
```diff
@@ -0,0 +1,123 @@
+`here <https://github.com/Vauxoo/addons-vauxoo/issues/new?body=module:%20
+default_warehouse_from_sale_team
+%0Aversion:%20
+14.0
```
15.0

rolandojduartem replied:
Done!

luisg123v commented on default_warehouse_from_sale_team/i18n/es.po:
```diff
@@ -0,0 +1,287 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Odoo Server 14.0+e\n"
```
15.0

rolandojduartem replied:
Done!

luisg123v commented on default_warehouse_from_sale_team/models/res_users.py:
```diff
@@ -0,0 +1,34 @@
+        """
+        user_wrong_team = self.filtered(lambda u: u.sale_team_id - u.sale_team_ids)
+        if user_wrong_team:
+            raise ValidationError(_(
```
Remove this constraint, it doesn't work as it should anymore, as it's triggered when adding a member to a sales team for the first time.

rolandojduartem replied:
Done!

luisg123v replied:
@rolandojduartem after furter investigation, I noticed the constraint works as expected, but the error message was wrong.

This was just fixed by @CarmenMiranda on #1531.

Could you include that commit, please?

rolandojduartem replied:
@luisg123v this commit was added, coul you review it, please?

luisg123v commented on default_warehouse_from_sale_team/models/default_warehouse_mixin.py:
```diff
@@ -0,0 +1,63 @@
+            ('model', '=', self._name),
+            ('relation', '=', 'stock.warehouse'),
+            ('ttype', '=', 'many2one'),
+            ('name', 'in', list(defaults)),
```
```suggestion
            ('name', 'in', fields_list),
```

Otherwise only already-set values will be considered

rolandojduartem replied:
Done!

luisg123v reviewed (changes requested):
Besides history, there should be only one commit for migration

luisg123v commented:
@rolandojduartem Remove task ID from commit's title, leave it only in PR's title.

luisg123v commented:
Hi @desdelinux and @rolandojduartem,

[Related task](https://www.vauxoo.com/web?debug=true#id=53868&action=470&active_id=7779&model=project.task&view_type=form&cids=1%2C12%2C14%2C16%2C17%2C18&menu_id=) says migrate stock by warehouse, that's module `stock_by_warehouse`. Are you sure you're migrating the right module?

desdelinux commented:
> Hi @desdelinux and @rolandojduartem,
> 
> [Related task](https://www.vauxoo.com/web?debug=true#id=53868&action=470&active_id=7779&model=project.task&view_type=form&cids=1%2C12%2C14%2C16%2C17%2C18&menu_id=) says migrate stock by warehouse, that's module `stock_by_warehouse`. Are you sure you're migrating the right module?

Yes, in addition to this, 2 modules related to the same task must be migrated

https://github.com/Vauxoo/addons-vauxoo/tree/14.0/stock_by_warehouse

https://github.com/Vauxoo/addons-vauxoo/tree/14.0/stock_by_warehouse_sale

luisg123v commented:
@desdelinux, Got it.

Remember also this one:
- https://github.com/Vauxoo/addons-vauxoo/tree/14.0/default_warehouse_from_sale_team_jit

luisg123v committed:
[FIX] default_warehouse_from_sale_team: typo in README

Which was causing list of contributors to not be correctly rendered.

rolandojduartem commented:
Hi @luisg123v I made the recommended changes and I left you a question about the dependency.

rolandojduartem commented:
Hi @luisg123v, I think you can review this pull request, please?

desdelinux commented:
@rolandojduartem Functional Review is ok.

@suniagajose your review, please.

cc  @hugho-ad @luisg123v

rolandojduartem committed:
[ADD] default_warehouse_from_sale_team: Allow to run procurement in warehouses not allowed on current user's sales teams

As part of migration to 15.0 of default_warehouse_from_sale_team and
default_warehouse_from_sale_team_jit, the dependency procurement_jit was removed
(https://github.com/odoo/odoo/commit/a838f2af683e03cd03ba02532f002ccd5c2642f5). This change caused
that default_warehouse_from_sale_team_jit
(https://github.com/Vauxoo/addons-vauxoo/tree/14.0/default_warehouse_from_sale_team_jit)
be deprecated on Odoo v15.0 and the main function of that model can be added in
default_warehouse_from_sale_team.

rolandojduartem commented:
Hi, @luisg123v  y @desdelinux, I added the functionality from default_warehouse_from_sale_team_jit in this module because procurement_jit was removed (details on PR description and commit), could you review this change, please?

MR: https://github.com/Vauxoo/addons-vauxoo/pull/1529
Dummy MR: https://git.vauxoo.com/vauxoo/villagroup/-/merge_requests/167

cc @hugho-ad

CarmenMiranda committed:
[FIX] default_warehouse_from_sale_team: Sale Team Constraint Message

Improving the message of the `sale_team_id` constraint, also, changing
the `string` and `help` of the `sale_team_ids` field because is not
readonly and the allowed teams should be set that field.

rolandojduartem committed:
[MIG] default_warehouse_from_sale_team: Migrate to 15.0

desdelinux commented:
Functional Review is Ok

@luisg123v @rolandojduartem

luisg123v committed:
[IMP] default_warehouse_from_sale_team: journal when created by code

When invoices are created by code (i.e., not functionally), journal is
never computed because it's currently set only by an onchange.

luisg123v committed:
[FIX] default_warehouse_from_sale_team: access error on journal name

Avoid access errors when user has access to a journal item but not to its
journal, e.g. when reconciling bank statements.

rolandojduartem commented:
Hi, @desdelinux, as I told you, three fixes were added on v14 and I added them to v15 in this pull request. Those changes are:
- https://github.com/Vauxoo/addons-vauxoo/pull/1531
- https://github.com/Vauxoo/addons-vauxoo/pull/1534
- https://github.com/Vauxoo/addons-vauxoo/pull/1535
So, may you do the functional review, please?
- Dummy MR: https://git.vauxoo.com/vauxoo/villagroup/-/merge_requests/167
Also, @luisg123v, may you do the technical review, please?

cc @hugho-ad

luisg123v reviewed (changes requested):
A lot of commits are missing. For instance:

- 0295bdd52 (you squashed it with cb9ccacf9)
- 144710b48 (you squashed it with a1270b8c9)


There should be 31 historical commits, but there are only 19 (your PR contains 21, which is 19 historical + two that are yours).

My suggestion is:
- Bring commits from v14 again
- Apply manually your commits by cherry-picking 4004e3f81 and 186edf8e2

rolandojduartem commented:
@luisg123v, done! Now, there are 31 historical and 2 new commits for migration to 15.0, may you review it, please?
